File: Facultate/Anul 2/Sem 2/MPP/Laburi/Lab3/JavaLab3/src/main/java/repos/ReservationsDBRepo.java
package repos;

import models.Reservation;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Properties;

public class ReservationsDBRepo implements ReservationsRepo {

    private final JdbcUtils dbUtils;
    private final Logger logger = LogManager.getLogger();

    public ReservationsDBRepo(Properties properties) {
        logger.info("Constructing ReservationsDBRepo with properties: {} ", properties);
        this.dbUtils = new JdbcUtils(properties);
    }

    /**
     * Finds a Reservation in the database by its ID
     * @param id the ID of the Reservation to find
     * @return an Optional containing the Reservation if found, or an empty Optional if not found or if there was an error accessing the database
     */
    @Override
    public Optional<Reservation> findOne(Long id) {
        String sql = "SELECT * FROM reservations WHERE id = ?";
        Connection connection = dbUtils.getConnection();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            ResultSet rs = ps.executeQuery();

            if (rs.next()) {
                Reservation reservation = reservationFromResultSet(rs);
                logger.info("Reservation found with id {}: {}", id, reservation);
                return Optional.of(reservation);
            } else {
                logger.info("No reservation found with id {}", id);
                return Optional.empty();
            }
        } catch (SQLException e) {
            logger.error(e);
            return Optional.empty();
        }
    }

    /**
     * Finds all Reservations in the database
     * @return a List of all Reservations found in the database, or an empty List if there was an error accessing the database
     */
    @Override
    public List<Reservation> findAll() {

        String sql = "SELECT * FROM reservations";
        List<Reservation> reservations = new ArrayList<>();
        Connection connection = dbUtils.getConnection();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ResultSet rs = ps.executeQuery();

            while (rs.next()) {
                Reservation reservation = reservationFromResultSet(rs);
                reservations.add(reservation);
            }

            logger.info("Found {} reservations in the database", reservations.size());

        } catch (SQLException e) {
            logger.error(e);
        }

        return reservations;
    }

    /**
     * Saves a new Reservation to the database
     * @param reservation the Reservation to save
     */
    @Override
    public void save(Reservation reservation) {

        String sql = "INSERT INTO reservations (client_name, reserved_seats_count, ride_id) VALUES (?, ?, ?)";
        Connection connection = dbUtils.getConnection();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, reservation.getClientName());
            ps.setInt(2, reservation.getReservedSeatsCount());
            ps.setLong(3, reservation.getRideId());
            ps.executeUpdate();
            logger.info("Reservation saved successfully: {}", reservation);

        } catch (SQLException e) {
            logger.error(e);
        }
    }

    /**
     * Deletes a Reservation from the database by its ID
     * @param id the ID of the Reservation to delete
     */
    @Override
    public void delete(Long id) {

        String sql = "DELETE FROM reservations WHERE id = ?";
        Connection connection = dbUtils.getConnection();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, id);
            ps.executeUpdate();
            logger.info("Reservation with id {} deleted successfully", id);

        } catch (SQLException e) {
            logger.error(e);
        }
    }

    /**
     * Updates an existing Reservation in the database
     * @param reservation the Reservation with updated data and the id of the Reservation to update
     */
    @Override
    public void update(Reservation reservation) {

        String sql = "UPDATE reservations SET client_name = ?, reserved_seats_count = ?, ride_id = ? WHERE id = ?";
        Connection connection = dbUtils.getConnection();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setString(1, reservation.getClientName());
            ps.setInt(2, reservation.getReservedSeatsCount());
            ps.setLong(3, reservation.getRideId());
            ps.setLong(4, reservation.getId());
            ps.executeUpdate();
            logger.info("Reservation updated successfully: {}", reservation);

        } catch (SQLException e) {
            logger.error(e);
        }
    }

    /**
     * Finds all Reservations in the database for a given ride ID
     * @param rideId the ID of the ride to find Reservations for
     * @return a List of Reservations for the given ride ID, or an empty List if there was an error accessing the database
     */
    @Override
    public List<Reservation> findReservationsByRideId(Long rideId) {

        String sql = "SELECT * FROM reservations WHERE ride_id = ?";
        List<Reservation> reservations = new ArrayList<>();
        Connection connection = dbUtils.getConnection();

        try (PreparedStatement ps = connection.prepareStatement(sql)) {
            ps.setLong(1, rideId);
            ResultSet rs = ps.executeQuery();

            while (rs.next()) {
                Reservation reservation = reservationFromResultSet(rs);
                reservations.add(reservation);
            }
            logger.info("Found {} reservations for ride id {}", reservations.size(), rideId);

        } catch (SQLException e) {
            logger.error(e);
        }

        return reservations;
    }

    /**
     * Helper method to create a Reservation object from a ResultSet
     * @param rs the ResultSet containing the data for a Reservation
     * @return a Reservation object created from the ResultSet
     * @throws SQLException if there was an error accessing the data in the ResultSet
     */
    private Reservation reservationFromResultSet(ResultSet rs) throws SQLException {

        String clientName = rs.getString("client_name");
        Integer reservedSeatsCount = rs.getInt("reserved_seats_count");
        Long rideId = rs.getLong("ride_id");

        Reservation reservation = new Reservation(clientName, reservedSeatsCount, rideId);
        reservation.setId(rs.getLong("id"));

        return reservation;
    }
}

File: Facultate/Anul 2/Sem 2/MPP/Laburi/Edi/CSharpServer/Server/repos/ReservationsDBRepo.cs
using MySql.Data.MySqlClient;
using NLog;
using Server.utils;
using Shared.models;

namespace Server.repos;

public class ReservationsDbRepo: IReservationsRepo
{
    private readonly DbUtils _dbUtils;
    private static readonly Logger Logger = LogManager.GetCurrentClassLogger();

    public ReservationsDbRepo(DbUtils dbUtils)
    {
        _dbUtils = dbUtils;
        Logger.Info("Reservations Repo created succesfully");
    }

    /// <summary>
    /// Finds a reservation by its ID
    /// </summary>
    /// <param name="id">The reservation's ID</param>
    /// <returns>The reservation with the specified ID or null if it isn't found</returns>
    public Reservation? FindOne(long id)
    {
        const string sql = "SELECT * FROM reservations WHERE id = @id";

        using var conn = _dbUtils.GetConnection();
        using var cmd = new MySqlCommand(sql, conn);
        cmd.Parameters.AddWithValue("@id", id);
        using var reader = cmd.ExecuteReader();
        if (reader.Read())
        {
            var reservation = ReservationFromReader(reader);
            Logger.Info("Reservation found: " + reservation);
            return reservation;
        }

        Logger.Info("Reservation not found: " + id);
        return null;
    }

    /// <summary>
    /// Finds all reservations in the database
    /// </summary>
    /// <returns>Anything that can be iterated on containing all reservations</returns>
    public IEnumerable<Reservation> FindAll()
    {
        const string sql = "SELECT * FROM reservations";

        var reservations = new List<Reservation>();
        using var conn = _dbUtils.GetConnection();
        using var cmd = new MySqlCommand(sql, conn);
        using var reader = cmd.ExecuteReader();

        while (reader.Read())
        {
            reservations.Add(ReservationFromReader(reader));
        }

        Logger.Info("Reservations found: " + reservations.Count);
        return reservations;
    }

    /// <summary>
    /// Finds all reservations for a ride in the database.
    /// </summary>
    /// <param name="rideId">The ride's ID</param>
    /// <returns>Anything that can be iterated on containing all reservations with the specified ride ID</returns>
    public IEnumerable<Reservation> FindReservationsByRide(long rideId)
    {
        const string sql = "SELECT * FROM reservations WHERE ride_id = @rideId";

        var reservations = new List<Reservation>();
        using var conn = _dbUtils.GetConnection();
        using var cmd = new MySqlCommand(sql, conn);
        cmd.Parameters.AddWithValue("@rideId", rideId);
        using var reader = cmd.ExecuteReader();

        while (reader.Read())
        {
            reservations.Add(ReservationFromReader(reader));
        }

        Logger.Info("Reservations found for ride " + rideId + ": " + reservations.Count);
        return reservations;
    }

    /// <summary>
    /// Saves a reservation to the database
    /// </summary>
    /// <param name="reservation">The reservation we are supposed to save</param>
    public void Save(Reservation reservation)
    {
        const string sql = "INSERT INTO reservations (client_name, reserved_seats_count, ride_id) VALUES (@clientName,@reservedSeatsCount,@rideId)";

        using var conn = _dbUtils.GetConnection();
        using var cmd = new MySqlCommand(sql, conn);

        cmd.Parameters.AddWithValue("@clientName", reservation.ClientName);
        cmd.Parameters.AddWithValue("@reservedSeatsCount", reservation.ReservedSeatsCount);
        cmd.Parameters.AddWithValue("@rideId", reservation.RideId);
        cmd.ExecuteNonQuery();
        Logger.Info("Reservation saved: Client Name = " + reservation.ClientName);
    }

    /// <summary>
    /// Deletes the reservation with the specified id
    /// </summary>
    /// <param name="id">The id of the reservation we want to delete</param>
    public void Delete(long id)
    {
        const string sql = "DELETE FROM reservations WHERE id = @id";

        using var conn = _dbUtils.GetConnection();
        using var cmd = new MySqlCommand(sql, conn);

        cmd.Parameters.AddWithValue("@id", id);
        cmd.ExecuteNonQuery();
        Logger.Info("Reservation with ID = " + id + "deleted");
    }

    /// <summary>
    /// Updates a reservation in the database. The reservation is identified by its id and all other fields are updated to the values of the given reservation
    /// </summary>
    /// <param name="reservation">The reservation with new data,and the id of the reservation we want to update</param>
    public void Update(Reservation reservation)
    {
        const string sql = "UPDATE reservations SET client_name = @clientName, reserved_seats_count = @reservedSeatsCount, ride_id = @rideId WHERE id = @id";

        using var conn = _dbUtils.GetConnection();
        using var cmd = new MySqlCommand(sql, conn);

        cmd.Parameters.AddWithValue("@id", reservation.Id);
        cmd.Parameters.AddWithValue("@clientName", reservation.ClientName);
        cmd.Parameters.AddWithValue("@reservedSeatsCount", reservation.ReservedSeatsCount);
        cmd.Parameters.AddWithValue("@rideId", reservation.RideId);
        cmd.ExecuteNonQuery();
        Logger.Info("Reservation updated: Client Name = " + reservation.ClientName);
    }

    /// <summary>
    /// Helper method that creates a Reservation object from a MySqlDataReader
    /// </summary>
    /// <param name="reader"> A MySqlDataReader containing all reservation information</param>
    /// <returns>The reservation that results from the MySqlDataReader</returns>
    private static Reservation ReservationFromReader(MySqlDataReader reader)
    {
        return new Reservation
        {
            Id = reader.GetInt64("id"),
            ClientName = reader.GetString("client_name"),
            ReservedSeatsCount = reader.GetInt32("reserved_seats_count"),
            RideId = reader.GetInt64("ride_id")
        };
    }
}

